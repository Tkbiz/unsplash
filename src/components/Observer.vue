<template>
  <div class="observer"/>
</template>

<script>
export default {
  props: ['options'],
  data: () => ({
    observer: null
  }),
  methods: {
    intersectionCallback(entries, observer) {
      entries.forEach(entry => {
        if (entry.isIntersecting) {
          this.$emit('intersect')
        }
      })
    }
  },
  mounted() {
    const options = this.options || {};
    this.observer = new IntersectionObserver(this.intersectionCallback, options)
    this.observer.observe(this.$el)
  },
  destroyed() {
    this.observer.disconnect()
  }
}
</script>
