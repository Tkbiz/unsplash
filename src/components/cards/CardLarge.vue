<template>
  <v-row v-if="card"
         class="card-large">
    <img @click.prevent="openModal()"
         class="col-12 col-sm-6 img-block"
         :src="card.urls.small ? card.urls.small : '@/assets/images/no-image.png'"
         alt="card image">
    <v-col
        cols="12"
        sm="6"
        class="py-0 custom-card">
      <div @click.prevent="openModal()"
           class="card-body">
        <div v-if="card.user.name"
             class="card-title">
          {{ card.user.name }}
        </div>
        <div v-if="card.alt_description" class="card-subtitle">
          {{ card.alt_description }}
        </div>
        <div v-if="card.description"
             class="card-text">
          {{card.description | truncate(140, '...')}}
        </div>
        <CardLargeCarousel
            :images="card.images"/>
        <div class="btn-group">
          <v-btn
              class="btn-order"
              outlined>
            order now
          </v-btn>
          <v-btn
              class="btn-menu"
              outlined>
            <v-icon>mdi-dots-horizontal</v-icon>
          </v-btn>
        </div>
      </div>
      <div class="card-footer">
        <div class="ml-lg-n6">
          <v-btn
              @click.prevent="downloadPhoto(card.urls.small)"
              icon>
            <v-icon>
              mdi-bookmark-outline
            </v-icon>
          </v-btn>
        </div>
        <div>
          <div>
          <span class="mr-5">
            <v-btn
                @click.prevent="setLike"
                icon>
              <v-icon>
                mdi-heart-outline
              </v-icon>
            </v-btn>
            <span class="numbers">
              {{ card.likes }}
            </span>
          </span>
            <span>
            <v-btn
                @click.prevent="setComment"
                icon>
              <v-icon>
                mdi-message-outline
              </v-icon>
            </v-btn>
            <span class="numbers">
              {{ card.downloads }}
            </span>
          </span>
          </div>
        </div>
      </div>
      <Modal
          v-if="modal"
          :photoStats="photoStats"
          :tagName="tagName"
          :samePhotos="samePhotos"
          :card="card"
          :userName="userName"
          :visibleModal="modal"
          @close="closeModal"/>
    </v-col>
  </v-row>
</template>

<script>
import CardLargeCarousel from '@/components/cards/CardLargeCarousel'
import Modal from '@/components/modal/Modal'
import {truncateTextMixin} from '@/mixins/truncateTextMixin'

export default {
  name: 'CardLarge',
  mixins: [truncateTextMixin],
  props: {
    card: {
      type: Object,
      required: true
    },
    userName: {
      required: true
    }
  },
  components: {
    CardLargeCarousel,
    Modal
  },
  data: () => ({
    modal: false,
    tagName: '',
    samePhotos: [],
    photoStats: {},
    page: 1
  }),
  methods: {
    async openModal() {
      this.modal = true
      this.card.tags.forEach(tag => this.tagName = tag.title)
      this.samePhotos = await this.$store.dispatch('searchPhotos', {
        tagName: this.tagName,
        page: this.page
      })
      this.photoStats = await this.$store.dispatch('fetchPhotoStats', this.card.id)
    },
    closeModal() {
      this.modal = false
    },
    setLike() {
      console.log('was liked!')
    },
    setComment() {
      console.log('comment was written!')
    },
    async downloadPhoto(url) {
      await this.$store.dispatch('downloadPhoto', url)
    }
  }
}
</script>

<style scoped>

</style>
