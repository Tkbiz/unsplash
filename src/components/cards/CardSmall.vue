<template>
  <v-col
      v-if="card"
      class="custom-card pa-0">
    <img @click.prevent="openModal(card)"
         class="card-img"
         :src="card.urls.small ? card.urls.small : '@/assets/images/no-image.png'"
         alt="card image">
    <div @click.prevent="openModal(card)"
         class="card-body">
      <div v-if="card.user.name"
           :class="{'mb-6': !card.user.location && !card.description}"
           class="card-title">
        {{ card.user.name }}
      </div>
      <div v-if="card.user.location"
           :class="{'mb-6': !card.description}"
           class="card-subtitle">
        {{ card.user.location }}
      </div>
      <div v-if="card.description"
           class="card-text">
        {{card.description | truncate(140, '...')}}
      </div>
    </div>
    <div class="card-footer">
      <div class="ml-n2">
        <v-btn
            @click.prevent="downloadPhoto(card.urls.small)"
            icon>
          <v-icon>
            mdi-bookmark-outline
          </v-icon>
        </v-btn>
      </div>
      <div>
        <div>
          <span class="mr-5">
            <v-btn
                @click.prevent="setLike"
                icon>
              <v-icon>
                mdi-heart-outline
              </v-icon>
            </v-btn>
            <span class="numbers">
              {{ card.likes }}
            </span>
          </span>
          <span>
            <v-btn
                @click.prevent="setComment"
                icon>
              <v-icon>
                mdi-message-outline
              </v-icon>
            </v-btn>
            <span class="numbers">
              {{ card.downloads }}
            </span>
          </span>
        </div>
      </div>
    </div>
    <Modal
        v-if="modal"
        :photoStats="photoStats"
        :tagName="tagName"
        :samePhotos="samePhotos"
        :card="card"
        :userName="userName"
        :visibleModal="modal"
        @close="closeModal"/>
  </v-col>
</template>

<script>
import Modal from '@/components/modal/Modal'
import {truncateTextMixin} from '@/mixins/truncateTextMixin'

export default {
  name: 'CardSmall',
  mixins: [truncateTextMixin],
  props: {
    card: {
      type: Object,
      required: true
    },
    userName: {
      required: true
    }
  },
  components: {
    Modal
  },
  data: () => ({
    modal: false,
    tagName: '',
    samePhotos: [],
    photoStats: {},
    page: 1
  }),
  methods: {
    async openModal() {
      this.modal = true
      this.card.tags.forEach(tag => this.tagName = tag.title)
      this.samePhotos = await this.$store.dispatch('searchPhotos', {
        tagName: this.tagName,
        page: this.page
      })
      this.photoStats = await this.$store.dispatch('fetchPhotoStats', this.card.id)
    },
    closeModal() {
      this.modal = false
    },
    setLike() {
      console.log('was liked!')
    },
    setComment() {
      console.log('comment was written!')
    },
    async downloadPhoto(url) {
      await this.$store.dispatch('downloadPhoto', url)
    }
  }
}
</script>

<style scoped>

</style>
