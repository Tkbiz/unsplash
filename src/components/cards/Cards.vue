<template>
  <v-container class="px-xl-0">
    <v-row class="mx-xl-n4">
      <v-col
          class="d-flex px-xl-4 py-0 mb-8"
          cols="12"
          sm="6"
          md="3"
          lg="3">
        <v-select
            class="custom-select"
            v-model="selectOrientation"
            @change="searchByOrientation"
            :items="items"
            label="Any orientation"
            outlined>
        </v-select>
      </v-col>
    </v-row>
    <v-row class="mx-xl-n4">
      <v-col
          cols="12"
          v-for="(card, index) of photosList"
          :key="card.id + index"
          class="d-flex px-xl-4 py-0 mb-8"
          :class="card.height < 4000 ? 'col-sm-6 col-lg-3': 'col-sm-12 col-lg-6'">
        <CardSmall
            v-if="card.height < 4000"
            :card="card"
            :userName="userName"/>
        <CardLarge
            v-else
            :card="card"
            :userName="userName"/>
      </v-col>
      <Loader v-if="loading"/>
      <Observer
          @intersect="showNewPhotos"
          :options="{root, threshold}"/>
    </v-row>
  </v-container>
</template>

<script>
import CardSmall from '@/components/cards/CardSmall'
import CardLarge from '@/components/cards/CardLarge'
import Observer from '@/components/Observer'
import {bus} from '@/bus'

export default {
  name: 'Cards',
  components: {
    CardSmall,
    CardLarge,
    Observer
  },
  props: {
    photosList: {
      type: Array,
      required: true
    },
    userName: {
      required: true
    }
  },
  data: () => ({
    page: 1,
    newPhotos: [],
    root: null,
    threshold: 0.5,
    loading: false,
    items: ['landscape', 'portrait', 'squarish'],
    selectOrientation: null
  }),
  methods: {
    async showNewPhotos() {
      this.loading = true
      this.page++
      this.newPhotos = await this.$store.dispatch('searchPhotos', {
        tagName: this.tagName,
        page: this.page
      })
      this.newPhotos.map(newPhoto => this.photosList.push(newPhoto))
      this.loading = false
    },
    async searchByOrientation() {
      await this.$store.dispatch('searchPhotos', {
        tagName: this.tagName,
        page: this.page,
        orientation: this.selectOrientation
      })
      bus.$emit('change-photos', this.photosListByOrientation)
    }
  },
  computed: {
    tagName() {
      return this.$store.getters.tagName
    },
    photosListByOrientation() {
      return this.$store.getters.photos
    }
  }
}
</script>
