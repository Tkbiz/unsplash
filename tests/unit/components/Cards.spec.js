import {mount, createLocalVue} from '@vue/test-utils'
import Vue from 'vue'
import Vuex from 'vuex'
import Vuetify from 'vuetify'
import Cards from '@/components/cards/Cards'
import CardSmall from '@/components/cards/CardSmall'
import CardLarge from '@/components/cards/CardLarge'
import Observer from '@/components/Observer'
import {bus} from '@/bus'

Vue.use(Vuetify)

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Cards.vue', () => {
  let vuetify
  let wrapper
  let actions
  let getters
  let store

  const observe = jest.fn()
  const unobserve = jest.fn()
  const disconnect = jest.fn()

  window.IntersectionObserver = jest.fn(() => ({
    observe,
    unobserve,
    disconnect
  }))

  beforeEach(() => {
    vuetify = new Vuetify()

    actions = {
      searchPhotos: jest.fn()
    }
    getters = {
      tagName: () => ('tag name'),
      photosListByOrientation: () => ([])
    }
    store = new Vuex.Store({
      getters,
      actions
    })

    wrapper = mount(Cards, {
      localVue,
      vuetify,
      store,
      propsData: {
        photosList: [],
        userName: {}
      },
      data: () => ({
        page: 1,
        newPhotos: [],
        root: null,
        threshold: 0.5,
        loading: false,
        newPhotosList: [],
        removablePhotosList: [],
        scrollableNewPhotos: [],
        items: ['landscape', 'portrait', 'squarish'],
        selectOrientation: null
      })
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy()
  })

  it("renders CardSmall", () => {
    expect(wrapper.exists(CardSmall)).toBe(true)
  })

  it("renders CardLarge", () => {
    expect(wrapper.exists(CardLarge)).toBe(true)
  })

  it("renders Observer", () => {
    expect(wrapper.exists(Observer)).toBe(true)
  })

  it("renders bus", () => {
    expect(wrapper.exists(bus)).toBe(true)
  })

  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders tagName using computed options', () => {
    const tagName = 'tag name'
    expect(JSON.stringify(wrapper.vm.tagName)).toBe(JSON.stringify(tagName))
  })

  it('when searchPhotos is called $emit is called', () => {
    wrapper.vm.$emit('change-photos', [])

    expect(wrapper.emitted('change-photos')).toBeTruthy()
    expect(wrapper.emitted('change-photos').length).toBe(1)
  })
})
