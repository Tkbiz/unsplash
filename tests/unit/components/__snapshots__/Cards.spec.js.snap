// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards.vue should match snapshot 1`] = `
<div class="container px-xl-0">
  <div class="row mx-xl-n4">
    <div class="d-flex px-xl-4 py-0 mb-8 col-sm-6 col-md-3 col-lg-3 col-12">
      <div class="v-input custom-select theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-55" class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span>â€‹</span></legend>
            </fieldset>
            <div class="v-select__slot"><label for="input-55" class="v-label theme--light" style="left: 0px; position: absolute;">Any orientation</label>
              <div class="v-select__selections"><input id="input-55" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
              </div><input type="hidden">
            </div>
            <div class="v-menu">
              <!---->
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="row mx-xl-n4">
    <!---->
    <div class="observer"></div>
  </div>
</div>
`;
